src/error.rs

```diff
@@ -0,0 +1,210 @@
+    impl std::convert::TryFrom<crate::ffi::QUIC_ERROR> for ErrorCode {
+      type Error = &'static str;
+      fn try_from(value: crate::ffi::QUIC_ERROR) -> Result<Self, Self::Error> {
+          match value {
+              $(
+                crate::ffi::$code1 => Ok(Self::$code1),
+              )*
+              _ => Err("Unknown QUIC_ERROR")
+          }
+      }
+    }
+  }
+}
+
+// defines all quic error codes.
+define_quic_error_code!(
+    QUIC_ERROR_SUCCESS,
+    QUIC_ERROR_PENDING,
+    QUIC_ERROR_CONTINUE,
+    QUIC_ERROR_OUT_OF_MEMORY,
+    QUIC_ERROR_INVALID_PARAMETER,
+    QUIC_ERROR_INVALID_STATE,
+    QUIC_ERROR_NOT_SUPPORTED,
+    QUIC_ERROR_NOT_FOUND,
+    QUIC_ERROR_BUFFER_TOO_SMALL,
+    QUIC_ERROR_HANDSHAKE_FAILURE,
+    QUIC_ERROR_ABORTED,
+    QUIC_ERROR_ADDRESS_IN_USE,
+    QUIC_ERROR_INVALID_ADDRESS,
+    QUIC_ERROR_CONNECTION_TIMEOUT,
+    QUIC_ERROR_CONNECTION_IDLE,
+    QUIC_ERROR_UNREACHABLE,
+    QUIC_ERROR_INTERNAL_ERROR,
+    QUIC_ERROR_CONNECTION_REFUSED,
+    QUIC_ERROR_PROTOCOL_ERROR,
+    QUIC_ERROR_VER_NEG_ERROR,
+    QUIC_ERROR_TLS_ERROR,
+    QUIC_ERROR_USER_CANCELED,
+    QUIC_ERROR_ALPN_NEG_FAILURE,
+    QUIC_ERROR_STREAM_LIMIT_REACHED,
+    QUIC_ERROR_ALPN_IN_USE,
+    QUIC_ERROR_CLOSE_NOTIFY,
+    QUIC_ERROR_BAD_CERTIFICATE,
+    QUIC_ERROR_UNSUPPORTED_CERTIFICATE,
+    QUIC_ERROR_REVOKED_CERTIFICATE,
+    QUIC_ERROR_EXPIRED_CERTIFICATE,
+    QUIC_ERROR_UNKNOWN_CERTIFICATE,
+    QUIC_ERROR_REQUIRED_CERTIFICATE,
+    QUIC_ERROR_CERT_EXPIRED,
+    QUIC_ERROR_CERT_UNTRUSTED_ROOT,
+    QUIC_ERROR_CERT_NO_CERT
+);
+
+impl From<ErrorCode> for QUIC_ERROR {
+    fn from(value: ErrorCode) -> Self {
+        value as QUIC_ERROR
+    }
+}
+
+/// The display string is the same as the debug string, i.e. the enum string.
+impl core::fmt::Display for ErrorCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        core::write!(f, "{:?}", self)
+    }
+}
+
+/// Quic error used in non-ffi code.
+/// Internal representation matches the os platfrom type.
+#[derive(Clone)]
+pub struct Error(pub QUIC_ERROR);
+
+impl Error {
+    /// Create an error from enum.
+    pub fn new(ec: ErrorCode) -> Self {
+        Self(ec as QUIC_ERROR)
+    }
+    /// Convert to error code if possible.
+    pub fn try_as_error_code(&self) -> Result<ErrorCode, &str> {
+        use std::convert::TryFrom;
+        ErrorCode::try_from(self.0 as QUIC_ERROR)
+    }
+
+    /// Convert from raw ffi error type.
+    pub fn from_raw(ec: QUIC_ERROR) -> Self {
+        Self(ec)
+    }
+
+    /// Temp api to use in manually written ffi code which is going to be
+    /// removed and replaced by auto generated ffi code.
+    /// This api will be replaced by from_raw.
+    pub fn from_u32(ec: u32) -> Self {
+        Self(ec as QUIC_ERROR)
+    }
+
+    /// Return Err if the error is considered a failure.
+    /// Ok includes both "no error" and "pending" status codes.
+    pub fn ok(self) -> Result<(), Self> {
```

Comment:
Just a question (I'm not that familiar with idiomatic error handling, so feel free to ignore)

Is it usual to use `ok` for converting an error type to a `Result`?
I am familiar with it to convert a `Result` to an `Option`, which basically means "it is ok to fail here". This is somewhat similar to what happens here ("this error is actually not an error").

But it leads to some strange patterns (`if let Err(e) = err.ok() { panic!(...)`). I am wondering if a different name could help, either for the method or for the type (`ResultCode` instead of `Error` since it is not always an error?).

Reply:
I need to add another struct type QuicError(pub QUIC_ERROR) and implement ok() on QuicError which returns the Error{} type. This maybe more confusing since we will have ErrorCode enum, Error struct, and QuicError struct. and it is hard to tell which to use.
I cannot add ok() on ErrorCode because it is not directly convertable from u32 and some u32 are not valid enums. I cannot add ok() to QUIC_ERROR because it is a type alias to u32/i32. 
U can create a windows_core::Error with S_OK. U might also able to create std::io::Error from os errno 0 which is success (not very sure).
So there are 2 options: 1. add QuicError struct 2. keep like this or make ok() pub(crate) or private.

Reply:
I agree with you, adding another type could be confusing.
I was talking at most about potentially renaming to avoid `ok`, which seems confusing to me. But I don't have a good name to offer, so please ignore this if you this the current version is good.

Reply:
I think in future PR I can change the ok() api to ok_from_raw(ec: QUIC_ERROR) -> Result<(),Error>) for dedicated conversion from ffi status returns. But that requires migrating to use ffi bindings that is still in my other branch.
I would like to merge the code as is and change it later. @nibanks